stdlib/public/Platform/android.modulemap

```diff
@@ -0,0 +1,677 @@
+  // Note: this module is not part of '_time'
+  // to prevent a circular modular dependency
+  // between linux_time and sys modules.
+  header "linux/time.h"
+  header "linux/time_types.h"
+  export *
+  explicit module bits_timespec {
+    header "bits/timespec.h"
+    export *
+  }
+}
+
+// Additional modules in the 'sys' subdirectory.
+module _sys_cdefs [system] {
+  header "sys/cdefs.h"
+  // Circular included header, so combine them
+  // into the same module.
+  header "android/api-level.h"
+  export *
+}
+module _sys_core [system] {
+  explicit module endian {
+    header "sys/endian.h"
+    export *
+  }
+  explicit module errno {
+    header "sys/errno.h"
+    export *
+  }
+  explicit module ifunc {
+    header "sys/ifunc.h"
+    export *
+  }
+  explicit module ipc {
+    header "sys/ipc.h"
+    export *
+  }
+  explicit module mman {
+    header "sys/mman.h"
+    export *
+  }
+  explicit module mman_common {
+    header "asm-generic/mman-common.h"
+    export *
+  }
+  explicit module msg {
+    header "sys/msg.h"
+    export *
+  }
+  explicit module random {
+    header "sys/random.h"
+    export *
+  }
+  explicit module resource {
+    header "sys/resource.h"
+    export *
+  }
+  explicit module sem {
+    header "sys/sem.h"
+    export *
+  }
+  explicit module shm {
+    header "sys/shm.h"
+    export *
+  }
+  explicit module un {
+    header "sys/un.h"
+    export *
+  }
+  explicit module utsname {
+    header "sys/utsname.h"
+    export *
+  }
+}
+module _sys_select [system] {
+  // Note: this module is not part of 
+  // 'sys_core' to prevent circular dependency error.
+  header "sys/select.h"
+  export *
+}
+// Use 'no_undeclared_includes' to avoid pulling in the libc++ module.
+// This module depends on 'stdint.h', which is defined in libc++.
+// We can't import libc++ as that would cause circular dependency
+// between libc++ and this module. Using 'no_undeclared_includes'
+// ensures that we include 'stdint.h' from usr/include instead of libc++.
+module _sys_types [system] [no_undeclared_includes] {
+  header "sys/types.h"
+  // The 'sys/types.h' header has a circular include
+  // with 'bits/pthread_types.h' and thus it must be in the same module.
+  header "bits/pthread_types.h"
+  export *
+  use _stdint
+  use _sys_cdefs
+  use _Builtin_stddef
+}
+// Use 'no_undeclared_includes' to avoid pulling in the libc++ module.
+// This module depends on 'stdint.h', which is defined in libc++.
+// We can't import libc++ as that would cause circular dependency
+// between libc++ and this module. Using 'no_undeclared_includes'
+// ensures that we include 'stdint.h' from usr/include instead of libc++.
+module _sys_user [system] [no_undeclared_includes]  {
+  header "sys/user.h"
+  export *
+  use _stdint
+  use _sys_cdefs
+  use _Builtin_stddef
+}
+
+// Module for zlib headers.
+module zlib [system] {
+  header "zlib.h"
+  explicit module zconf {
+    header "zconf.h"
+    export *
+  }
+  export *
+  link "z"
```

Comment:
One thing I noticed is that this line means that all libraries and executables built with this Android overlay are now newly linked against `libz.so`, despite basically none of them needing it:
```
> llvm-readelf -d usr/lib/swift/android/libFoundation.so | ag NEEDED
  0x0000000000000001 (NEEDED)       Shared library: [libswiftDispatch.so]
  0x0000000000000001 (NEEDED)       Shared library: [libicuuc.so.74]
  0x0000000000000001 (NEEDED)       Shared library: [libicui18n.so.74]
  0x0000000000000001 (NEEDED)       Shared library: [libicudata.so.74]
  0x0000000000000001 (NEEDED)       Shared library: [libdispatch.so]
  0x0000000000000001 (NEEDED)       Shared library: [libBlocksRuntime.so]
  0x0000000000000001 (NEEDED)       Shared library: [libswiftAndroid.so]
  0x0000000000000001 (NEEDED)       Shared library: [libswiftCore.so]
  0x0000000000000001 (NEEDED)       Shared library: [libswift_Concurrency.so]
  0x0000000000000001 (NEEDED)       Shared library: [libswift_StringProcessing.so]
  0x0000000000000001 (NEEDED)       Shared library: [libswift_RegexParser.so]
  0x0000000000000001 (NEEDED)       Shared library: [libm.so]
  0x0000000000000001 (NEEDED)       Shared library: [libz.so.1]
  0x0000000000000001 (NEEDED)       Shared library: [liblog.so]
  0x0000000000000001 (NEEDED)       Shared library: [libdl.so]
  0x0000000000000001 (NEEDED)       Shared library: [libc.so]
```

What is the great need to add this to the Android overlay? Packages that need this can simply add a separate zlib modulemap.

Reply:
Hmm this is an interesting point. I suppose they don't really need to, but we would have to take out the zlib headers from the 'Android' module to avoid that. That seems reasonable to me, as users can just include 'zlib' module directly from Swift then. Let me try to see if I can take the zlib headers from the 'Android' module and rerun the tests.

Reply:
> I removed the 'zlib' header inclusion, and updated the PR, so now zlib link dependency shouldn't be pulled in implicitly.

Looks like you made a mistake? This zlib link is still there in your latest push yesterday.

Reply:
Oh, I see now that you simply removed the header from the `SwiftAndroid` module. I can confirm that this fixes this problem in my latest build.